src/hist/plot.py

```diff
@@ -212,6 +220,309 @@ def plot2d_full(
+# TODO: Revise plot_ratio to make it possible for plot_pull to use as infrastructure
+def plot_ratio(
+    self: hist.BaseHist,
+    other: Union[hist.BaseHist, Callable[[np.ndarray], np.ndarray], str],
+    likelihood: bool = False,
+    *,
+    ax_dict: "Optional[Dict[str, matplotlib.axes.Axes]]" = None,
+    **kwargs: Any,
+) -> "Tuple[matplotlib.axes.Axes, matplotlib.axes.Axes]":
+    """
+    Plot_ratio method for BaseHist object.
+    """
+
+    try:
+        from iminuit import Minuit  # noqa: F401
+        from scipy.optimize import curve_fit  # noqa: F401
+    except ImportError:
+        print(
+            "Hist.plot_pull requires scipy and iminuit. Please install hist[plot] or manually install dependencies.",
+            file=sys.stderr,
+        )
+        raise
+
+    if self.ndim != 1:
+        raise TypeError(
+            f"Only 1D-histogram supports ratio plot, try projecting {self.__class__.__name__} to 1D"
+        )
+    if isinstance(other, hist.hist.Hist):
+        if other.ndim != 1:
+            raise TypeError(
+                f"Only 1D-histogram supports ratio plot, try projecting {other.__class__.__name__} to 1D"
+            )
+
+    if ax_dict:
+        try:
+            main_ax = ax_dict["main_ax"]
+            ratio_ax = ax_dict["ratio_ax"]
+        except KeyError:
+            raise ValueError("All axes should be all given or none at all")
+    else:
+        fig = plt.gcf()
+        grid = fig.add_gridspec(2, 1, hspace=0, height_ratios=[3, 1])
+
+        main_ax = fig.add_subplot(grid[0])
+        ratio_ax = fig.add_subplot(grid[1], sharex=main_ax)
+
+    # Keyword Argument Conversion: convert the kwargs to several independent args
+
+    # error bar keyword arguments
+    eb_kwargs = _filter_dict(kwargs, "eb_")
+    eb_kwargs.setdefault("label", "Histogram Data")
+
+    # fit plot keyword arguments
+    fp_kwargs = _filter_dict(kwargs, "fp_")
+    fp_kwargs.setdefault("label", "Fitted Value")
+
+    # uncertainty band keyword arguments
+    ub_kwargs = _filter_dict(kwargs, "ub_")
+    ub_kwargs.setdefault("label", "Uncertainty")
+
+    # # patch plot keyword arguments
+    # pp_kwargs = _filter_dict(kwargs, "pp_", ignore={"pp_num"})
+    # pp_num = kwargs.pop("pp_num", 5)
+
+    # Judge whether some arguments are left
+    if kwargs:
+        raise ValueError(f"{set(kwargs)}' not needed")
+
+    # Computation and Fit
+    x_values = self.axes[0].centers
+    numerator = self.values()
+
+    if callable(other) or type(other) in [str]:
+        variances = self.variances()
+        if variances is None:
+            raise RuntimeError(
+                "Cannot compute from a variance-less histogram, try a Weight storage"
+            )
+        numerator_uncert = np.sqrt(variances)
+
+        # Infer model parameters using `other` as fit model
+        popt, pcov = _curve_fit_wrapper(
+            other, x_values, numerator, numerator_uncert, likelihood=likelihood
+        )
```

Comment:
Question about types: If I run pre-commit I get 

```
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

src/hist/plot.py:403: error: Argument 1 to "_curve_fit_wrapper" has incompatible type "Union[BaseHist, Callable[[ndarray], ndarray], str]"; expected "Callable[..., Any]"
src/hist/plot.py:405: error: "BaseHist" not callable
src/hist/plot.py:405: error: "str" not callable
src/hist/plot.py:410: error: "BaseHist" not callable
src/hist/plot.py:410: error: "str" not callable
src/hist/plot.py:417: error: Item "str" of "Union[BaseHist, str]" has no attribute "values"
Found 6 errors in 1 file (checked 17 source files)
```

but I don't understand why I'm getting these error given the control flow (if this code is being executed then `other` is a callable and not a `hist.Basehist`). This need to get refactored anyway, but to get it to stop yelling at me Is it not possible to have type dependent control flow if you're using type checking in Python?

Reply:
Never use `type(other) in [str]`. Don't even use `type(other) in {str}`. Never directly compare types, always use `isinstance(other, str)`. Some flake8 checks will force this, I think.  This is why mypy can't type narrow on str as well, but not the only reason. It's perfectly valid to subclass str, but your version would break. For callable, it probably can't distinguish that callable will filter out BaseHist (or it might be fixed by the above fix). You can check `isinstance(other, BaseHist)`, then the narrowing should work.